src/runtime/cuda/nccl_api.cu

```diff
@@ -0,0 +1,832 @@
+    return std::pair<IdArray, NDArray>(in_idx, in_value);
+  }
+
+  std::pair<IdArray, NDArray> part_perm = part->GeneratePermutation(in_idx);
+  const IdType * const perm = static_cast<const IdType*>(part_perm.first->data);
+  const int64_t * const send_sum =
+      static_cast<const int64_t*>(part_perm.second->data);
+
+  IdType * send_idx = static_cast<IdType*>(device->AllocWorkspace(ctx,
+      num_in*sizeof(IdType)));
+  DType * send_value = static_cast<DType*>(device->AllocWorkspace(ctx,
+      num_in*num_feat*sizeof(DType)));
+
+  // permute the indices and values
+  {
+    const dim3 block(256);
+    const dim3 grid((num_in+block.x-1)/block.x);
+
+    _DualPermKernel<<<grid, block, 0, stream>>>(
+        static_cast<const IdType*>(in_idx->data),
+        static_cast<const DType*>(in_value->data),
+        perm,
+        num_in,
+        num_feat,
+        send_idx,
+        send_value);
+    CUDA_CALL(cudaGetLastError());
+  }
+
+  // compute the prefix sum of the send values
+  int64_t * send_prefix = static_cast<int64_t*>(
+      device->AllocWorkspace(ctx, sizeof(int64_t)*(comm_size+1)));
+  {
+    size_t prefix_workspace_size;
+    CUDA_CALL(cub::DeviceScan::ExclusiveSum(nullptr, prefix_workspace_size,
+        send_sum, send_prefix, comm_size+1, stream));
+
+    void * prefix_workspace = device->AllocWorkspace(
+        ctx, prefix_workspace_size);
+    CUDA_CALL(cub::DeviceScan::ExclusiveSum(prefix_workspace, prefix_workspace_size,
+        send_sum, send_prefix, comm_size+1, stream));
+    device->FreeWorkspace(ctx, prefix_workspace);
+  }
+
+  std::vector<int64_t> send_prefix_host(comm_size+1);
+  device->CopyDataFromTo(
+      send_prefix,
+      0,
+      send_prefix_host.data(),
+      0,
+      send_prefix_host.size()*sizeof(*send_prefix),
+      ctx,
+      DGLContext{kDLCPU, 0},
+      DGLType{kDLInt, sizeof(*send_prefix)*8, 1},
+      stream);
+  device->FreeWorkspace(ctx, send_prefix);
+
+  CHECK_EQ(send_prefix_host.back(), num_in);
+
+  // communicate the amount to send
+  int64_t * recv_sum = static_cast<int64_t*>(
+      device->AllocWorkspace(ctx, sizeof(int64_t)*(comm_size+1)));
+  comm->AllToAll(send_sum, recv_sum, 1, stream);
+
+  // compute the prefix sum of the recv values
+  int64_t * recv_prefix = static_cast<int64_t*>(
+      device->AllocWorkspace(ctx, sizeof(int64_t)*(comm_size+1)));
+  {
+    size_t prefix_workspace_size;
+    CUDA_CALL(cub::DeviceScan::ExclusiveSum(nullptr, prefix_workspace_size,
+        recv_sum, recv_prefix, comm_size+1));
+
+    void * prefix_workspace = device->AllocWorkspace(
+        ctx, prefix_workspace_size);
+    CUDA_CALL(cub::DeviceScan::ExclusiveSum(prefix_workspace, prefix_workspace_size,
+        recv_sum, recv_prefix, comm_size+1));
+    device->FreeWorkspace(ctx, prefix_workspace);
+  }
+  device->FreeWorkspace(ctx, recv_sum);
+
+  // finally copy the prefixsum sum down to the host
+  std::vector<int64_t> recv_prefix_host(comm_size+1);
```

Comment:
Do we need to copy recv_predix into CPU?

Reply:
This is needed for the call to NCCL, to specify which portions of the array need to be sent to which processors. See https://github.com/dmlc/dgl/pull/2825/files#diff-5d6e33a4f7ce35295db947aff3e3cb6462f630d6a7b85629c3b2a3eb5c32e98dR82